plugins/soundsourcemediafoundation/audiosourcemediafoundation.cpp

```diff
@@ -0,0 +1,558 @@
+#include "audiosourcemediafoundation.h"
+
+#include <mfapi.h>
+#include <mfidl.h>
+#include <mfreadwrite.h>
+#include <mferror.h>
+#include <propvarutil.h>
+
+#include <QtDebug>
+
+namespace Mixxx {
+
+namespace
+{
+
+const bool sDebug = false;
+
+const int kNumChannels = 2;
+const int kSampleRate = 44100;
+const int kLeftoverSize = 4096; // in sample_type's, this seems to be the size MF AAC
+const int kBitsPerSampleForBitrate = 16; // for bitrate calculation decoder likes to give
+
+/**
+ * Convert a 100ns Media Foundation value to a number of seconds.
+ */
+inline qreal secondsFromMF(qint64 mf) {
+    return static_cast<qreal>(mf) / 1e7;
+}
+
+/**
+ * Convert a number of seconds to a 100ns Media Foundation value.
+ */
+inline qint64 mfFromSeconds(qreal sec) {
+    return sec * 1e7;
+}
+
+/**
+ * Convert a 100ns Media Foundation value to a frame offset.
+ */
+inline qint64 frameFromMF(qint64 mf) {
+    return static_cast<qreal>(mf) * kSampleRate / 1e7;
+}
+
+/**
+ * Convert a frame offset to a 100ns Media Foundation value.
+ */
+inline qint64 mfFromFrame(qint64 frame) {
+    return static_cast<qreal>(frame) / kSampleRate * 1e7;
+}
+
+/** Microsoft examples use this snippet often. */
+template<class T> static void safeRelease(T **ppT) {
+    if (*ppT) {
+        (*ppT)->Release();
+        *ppT = NULL;
+    }
+}
+
+}
+
+AudioSourceMediaFoundation::AudioSourceMediaFoundation()
+        : m_hrCoInitialize(E_FAIL), m_hrMFStartup(E_FAIL),
+                m_pReader(NULL), m_pAudioType(NULL), m_wcFilename(
+                NULL), m_nextFrame(0), m_leftoverBuffer(NULL), m_leftoverBufferSize(
+                0), m_leftoverBufferLength(0), m_leftoverBufferPosition(0), m_mfDuration(
+                0), m_iCurrentPosition(0), m_dead(false), m_seeking(false) {
+
+    // these are always the same, might as well just stick them here
+    // -bkgood
+    // AudioSource properties
+    setChannelCount(kNumChannels);
+    setFrameRate(kSampleRate);
+
+    // presentation attribute MF_PD_AUDIO_ENCODING_BITRATE only exists for
+    // presentation descriptors, one of which MFSourceReader is not.
+    // Therefore, we calculate it ourselves, assuming 16 bits per sample
+    setBitrate(frames2samples(kBitsPerSampleForBitrate * kSampleRate) / 1000);
+}
+
+AudioSourceMediaFoundation::~AudioSourceMediaFoundation() {
+    close();
+}
+
+AudioSourcePointer AudioSourceMediaFoundation::open(QString fileName) {
+    AudioSourceMediaFoundation* pAudioSourceMediaFoundation(new AudioSourceMediaFoundation);
```

Comment:
Please note that the plain pointer always has the type of the sub class whereas the managed pointer has the type of the super class. The alternative would be to use an explicit downcast to invoke the postConstruct() method. Downcasts are code smell ;)

I would instead prefer to align the signature of the postConstruct() method and move it as a pure virtual function into the base class. Together with another factory method for the construction of sub class instances this could lead the way for also moving the open() method into the super class.